src/connectors/impls/gcs/api_client.rs

```diff
@@ -0,0 +1,476 @@
+            }
+        } else {
+            return Err("no response from GCS".into());
+        }
+
+        Ok(())
+    }
+
+    async fn start_upload(&mut self, url: &Url<HttpsDefaults>, file: FileId) -> Result<()> {
+        for i in 0..3 {
+            let url = url::Url::parse(&format!(
+                "{}/b/{}/o?name={}&uploadType=resumable",
+                url, file.bucket, file.name
+            ))?;
+            #[cfg(not(test))]
+            let mut request = Request::new(Method::Post, url);
+            #[cfg(test)]
+            let request = Request::new(Method::Post, url);
+            #[cfg(not(test))]
+            {
+                request.insert_header("Authorization", self.token.header_value()?.to_string());
+            }
+
+            let response = self.client.send(request).await;
+
+            if let Ok(mut response) = response {
+                if !response.status().is_server_error() {
+                    self.sessions_per_file.insert(
+                        file,
+                        url::Url::parse(
+                            response
+                                .header("Location")
+                                .ok_or(ErrorKind::GoogleCloudStorageError(
+                                    "Missing Location header",
+                                ))?
+                                .get(0)
+                                .ok_or(ErrorKind::GoogleCloudStorageError(
+                                    "Missing Location header value",
+                                ))?
+                                .as_str(),
+                        )?,
+                    );
+
+                    break;
+                }
+
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+            }
+
+            sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+            continue;
+        }
+
+        Ok(())
+    }
+
+    async fn finish_upload(&mut self, file: FileId, data: BufferPart) -> Result<()> {
+        let session_url: url::Url =
+            self.sessions_per_file
+                .remove(&file)
+                .ok_or(ErrorKind::GoogleCloudStorageError(
+                    "No session URL is available",
+                ))?;
+
+        for i in 0..3 {
+            let mut request = Request::new(Method::Put, session_url.clone());
+
+            request.insert_header(
+                "Content-Range",
+                format!(
+                    "bytes {}-{}/{}",
+                    data.start,
+                    // -1 on the end is here, because Content-Range is inclusive
+                    data.start + data.len() - 1,
+                    data.start + data.len()
+                ),
+            );
+
+            request.set_body(data.data.clone());
+
+            let response = self.client.send(request).await;
+
+            if let Ok(mut response) = response {
```

Comment:
Same as above, it'd be nice not to swallow this error completely if something goes wrong.